bitcoin/src/taproot.rs

```diff
@@ -521,7 +555,188 @@ impl Default for TaprootBuilder {
+            IncompleteBuilder::NotFinalized(builder) | IncompleteBuilder::HiddenParts(builder) =>
+                builder,
+        }
+    }
+}
+
+impl core::fmt::Display for IncompleteBuilder {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.write_str(match self {
+            IncompleteBuilder::NotFinalized(_) =>
+                "an attempt to construct a tap tree from a builder containing incomplete branches.",
+            IncompleteBuilder::HiddenParts(_) =>
+                "an attempt to construct a tap tree from a builder containing hidden parts.",
+        })
+    }
+}
+
+#[cfg(feature = "std")]
+#[cfg_attr(docsrs, doc(cfg(feature = "std")))]
+impl std::error::Error for IncompleteBuilder {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        use self::IncompleteBuilder::*;
+
+        match self {
+            NotFinalized(_) | HiddenParts(_) => None,
+        }
+    }
+}
+
+/// Error happening when [`TapTree`] is constructed from a [`NodeInfo`]
+/// having hidden branches.
+#[derive(Clone, Ord, PartialOrd, Eq, PartialEq, Hash, Debug)]
+#[non_exhaustive]
+pub enum HiddenNodes {
+    /// Indicates an attempt to construct a tap tree from a builder containing hidden parts.
+    HiddenParts(NodeInfo),
+}
+
+impl HiddenNodes {
+    /// Converts error into the original incomplete [`NodeInfo`] instance.
+    pub fn into_builder(self) -> NodeInfo {
+        match self {
+            HiddenNodes::HiddenParts(node_info) => node_info,
+        }
+    }
+}
+
+impl core::fmt::Display for HiddenNodes {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.write_str(match self {
+            HiddenNodes::HiddenParts(_) =>
+                "an attempt to construct a tap tree from a node_info containing hidden parts.",
+        })
+    }
+}
+
+#[cfg(feature = "std")]
+#[cfg_attr(docsrs, doc(cfg(feature = "std")))]
+impl std::error::Error for HiddenNodes {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        use self::HiddenNodes::*;
+
+        match self {
+            HiddenParts(_) => None,
+        }
+    }
+}
+
+/// Taproot Tree representing a complete binary tree without any hidden nodes.
+/// This is in contrast to [`NodeInfo`], which allows hidden nodes.
+//
+// This is a bug in BIP370 that does not specify how to share trees with hidden nodes,
+// for which we need a separate type.
+#[derive(Clone, Debug)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+#[cfg_attr(feature = "serde", serde(into = "NodeInfo"))]
+#[cfg_attr(feature = "serde", serde(try_from = "NodeInfo"))]
+pub struct TapTree(pub(crate) NodeInfo);
+
+impl PartialEq for TapTree {
+    fn eq(&self, other: &Self) -> bool { self.0.hash.eq(&other.0.hash) }
+}
+
+impl core::hash::Hash for TapTree {
+    fn hash<H: core::hash::Hasher>(&self, state: &mut H) { self.0.hash(state) }
+}
```

Comment:
Perhaps, I should add some documentation to this. For taproot trees, because the leaves get sorted by hash while constructing the root, a tree with root A, left child B and right child C is the same as root A with left child B and right child A. 

We had `Eq` reflect this property in `TapTree`, so I thought `Hash` should reflect this property. The `Eq` property is really handy while round-tripping psbts.

Reply:
If `Eq` returns `true` then comparing hashes *must* also return `true`.

I'm confused by this explanation:

> a tree with root A, left child B and right child C is the same as root A with left child B and right child A.

Maybe you meant to say this?

> a tree with root A, left child B and right child C is the same as root A with left child C and right child B.